### PR TITLE
Flp 107 earnings recalculation

### PR DESCRIPTION
**Pull request by tom-gough**

### CODE REVIEW

File: src/AcceptanceTests/Helpers/ApprenticeshipEntityExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+﻿using Microsoft.Extensions.Internal;
+using SFA.DAS.Funding.ApprenticeshipEarnings.Domain.Services;
+using SFA.DAS.Funding.ApprenticeshipEarnings.DurableEntities;
+using SFA.DAS.Funding.ApprenticeshipEarnings.DurableEntities.Models;
+
+namespace SFA.DAS.Funding.ApprenticeshipEarnings.AcceptanceTests.Helpers;
+
+internal static class ApprenticeshipEntityExtensions
+{
+    internal static ApprenticeshipEpisodeModel GetCurrentEpisode(this ApprenticeshipEntity apprenticeship, ISystemClockService systemClock)
```
Comment (MikeV-TalentConsulting) [inline src/AcceptanceTests/Helpers/ApprenticeshipEntityExtensions.cs:10]: There are 2 test files for getCurrentEpisode which are exactly the same and a very similar one in the main code, could we reduce the code duplication

File: src/Domain/Apprenticeship/Apprenticeship.cs [hunk 2]
```diff
@@ -1,87 +1,51 @@
...
-        var newEarnings = apprenticeshipFunding.RecalculateEarnings(existingEarnings, effectiveFromDate);
-        EarningsProfile = new EarningsProfile(apprenticeshipFunding.OnProgramTotal, newEarnings.Select(x => new Instalment(x.AcademicYear, x.DeliveryPeriod, x.Amount)).ToList(), apprenticeshipFunding.CompletionPayment, Guid.NewGuid());
-        AddEvent(new EarningsRecalculatedEvent(this));
-    }
+        var episode = this.ApprenticeshipEpisodes.Single(x => x.ApprenticeshipEpisodeKey == episodeUpdatedEvent.Episode.Key);
+        episode.Update(episodeUpdatedEvent.Episode);
+
+        if (episodeUpdatedEvent is ApprenticeshipPriceChangedEvent apprenticeshipPriceChangedEvent)
+        {
+            var existingEarnings = episode.EarningsProfile.Instalments.Select(x => new Earning { AcademicYear = x.AcademicYear, Amount = x.Amount, DeliveryPeriod = x.DeliveryPeriod }).ToList();
```
Comment (MikeV-TalentConsulting) [inline src/Domain/Apprenticeship/Apprenticeship.cs:41]: nullable warning

File: src/AcceptanceTests/Helpers/ApprenticeshipEntityExtensions.cs (hunk 1 above)
Comment (tom-gough) [reply src/AcceptanceTests/Helpers/ApprenticeshipEntityExtensions.cs:10]: sorted

File: src/Domain/Apprenticeship/Apprenticeship.cs (hunk 2 above)
Comment (tom-gough) [reply src/Domain/Apprenticeship/Apprenticeship.cs:41]: sorted

File: src/Domain.UnitTests/ApprenticeshipEpisode/WhenUpdateAgreedPrice.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        BuildApprenticeshipEpisode();
+
+        // Act
+        _apprenticeshipEpisode.UpdateAgreedPrice(_clockService.Object, _newAgreedPrice, _deletedPriceKeys, _newPriceKey);
+
+        // Assert
+        _apprenticeshipEpisode.Prices.Should().NotContain(p => _deletedPriceKeys.Contains(p.PriceKey));
+    }
+
+    //todo ThenTheExistingActivePriceIsClosedOff
```
[the remark this answers is not in the document]
Comment (tom-gough) [reply src/Domain.UnitTests/ApprenticeshipEpisode/WhenUpdateAgreedPrice.cs:64]: sorted

File: src/Domain.UnitTests/TestHelpers/FixtureExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        {
+            UKPRN = x.UKPRN,
+            EmployerAccountId = x.EmployerAccountId,
+            AgeAtStartOfApprenticeship = x.AgeAtStartOfApprenticeship,
+            TrainingCode = x.TrainingCode,
+            FundingType = x.FundingType,
+            LegalEntityName = x.LegalEntityName,
+            EarningsProfile = withMissingEarningsProfile ? null : MapEarningsProfileToModel(x.EarningsProfile),
+            FundingEmployerAccountId = x.FundingEmployerAccountId,
+            Prices = MapPricesToModel(x.Prices, newPrice == null ? apprenticeship.ApprenticeshipEpisodes.Single().Prices.First().AgreedPrice + 1 : newPrice.Value + 1, newStartDate), //todo is .First() right here?
```
[the remark this answers is not in the document]
Comment (tom-gough) [reply src/Domain.UnitTests/TestHelpers/FixtureExtensions.cs:58]: I've changed this to .Single there should only be one test price in this scenario

**Pull request closed without merging**